### DISCUSSION TITLE
How to open a URL from a plugin?

### DISCUSSION THREAD

**Discussion by David-Maisonave** [stashapp/stash#5418] [Q&A]
How can I open a URL from a plugin?
I created plugin [DupFileManager](https://github.com/David-Maisonave/Axter-Stash/blob/main/plugins/DupFileManager).

It creates a report similar to that of Scene Tools "Duplicate Checker".
It creates multiple HTML pages, and I'm trying to figure out a way to have it display the page after the plugin task is complete.
I tried using the following code, but 
#### Code#1
`os.system(f"start file:///{htmlReportName}")`

#### Code#2
```
import webbrowser
webbrowser.open(htmlReportName, new=2, autoraise=True)
```

Neither version works. When I run this, nothing happens.

Is there any way for a plugin to display a link on stash, that  is besides the one link in the Plugins menu?

**Comment by DogmaDragon** [comment]
From [Discord](https://discord.com/channels/559159668438728723/742220889017417738/1298812926546677783):
> python cannot access the browser, in the same way the browser cannot access python
>
> you need a ui plugin, using stash as the rpc connector via logging

**Comment by feederbox826** [comment]
I replied directly on discord, both sides are sandboxed, you might want to do something like [tagGraph](https://github.com/stg-annon/StashScripts/tree/main/plugins/tagGraph) and always output to the same file

**Comment by David-Maisonave** [comment]
> you need a ui plugin, using stash as the rpc connector via logging

Are there any python UI plugin examples?
I looked at the following link [uipluginapi](https://docs.stashapp.cc/in-app-manual/plugins/uipluginapi/), and it refers to react-component example.
I would like to be able to refer to a second example.

**Reply by feederbox826** [reply]
There is no Python UI since it cannot run in the browser. You need two separate parts to interact with each other.

Here is an example of escaping sandbox from UI -> Python https://github.com/7dJx1qP/stash-plugins/tree/main/plugins/stashOpenMediaPlayer

[Stash Helper](https://github.com/philpw99/Stash_Helper) escape sandbox from UI -> Python

[server](https://github.com/feederbox826/StashOpen) escapes UI sandbox by listening locally with a [UI Plugin](https://github.com/feederbox826/plugins/tree/main/plugins/stash-open)

**Comment by David-Maisonave** [comment]
I found some other UI plugin examples, but I still don't have a clue how to use it.
The stash document is not very clear. I wish they had a Hello World version, or a UI plugin for dummies.

**Comment by David-Maisonave** [comment]
In the Stash source code, UI plugin example react-component has a [testReact.yml](https://github.com/stashapp/stash/blob/develop/pkg/plugin/examples/react-component/src/testReact.yml), and that file refers to testReact.js, but there is no file with that name in the source code.
There is a file called testReact.tsx instead.

Is that an error in the testReact.yml source code, and should it be testReact.tsx instead?

**Reply by feederbox826** [reply]
No, it has to be compiled

**Comment by David-Maisonave** [comment]
> No, it has to be compiled

Thanks. I just compiled, and I copied the build files to the plugin folder.
`C:\Users\MyUserName\.stash\plugins\testReact`

But it still didn't work.  I tried the following link, and it gives me a **"Not Found"** error.
http://localhost:9999/plugin/test-react

Also the tools tab [http://localhost:9999/settings?tab=tools](http://localhost:9999/settings?tab=tools) is supposed to have **test-react** listed, but it doesn't.

I'm going to close this question, and open a new question to specifically ask help in getting testReact working.

**Comment by David-Maisonave** [comment]
FYI:
I was able to get the testReact UI plugin to work.  I'm not sure why, but UI plugins have a delay before they show up under the tools menu.
